optuna/integration/mlflow.py

```diff
@@ -102,56 +141,124 @@ def __init__(
+        """
+
+        def decorator(func: ObjectiveFuncType) -> ObjectiveFuncType:
+            def wrapper(trial: optuna.trial.Trial) -> Union[float, Sequence[float]]:
+                study = trial.study
+                self._initialize_experiment(study)
+
+                with mlflow.start_run(run_name=str(trial.number), nested=self._nest_trials) as run:
+                    trial.set_user_attr(RUN_ID_ATTRIBUTE_KEY, run.info.run_id)
+
+                    return func(trial)
+
+            return wrapper
+
+        return decorator
+
+    def _initialize_experiment(self, study: optuna.study.Study) -> None:
+        """Initialize a MLFlow experiment with the study name.
+
+        If a tracking uri has been provided, MLFlow will be initialized to use it.
+
+        Args:
+            study: Study to be tracked in MLFlow
+        """
+        # This sets the tracking_uri for MLflow.
+        if self._tracking_uri is not None:
+            mlflow.set_tracking_uri(self._tracking_uri)
+
+        # This sets the experiment of MLflow.
+        mlflow.set_experiment(study.study_name)
+
+    def _set_tags(self, trial: optuna.trial.FrozenTrial, study: optuna.study.Study) -> None:
+        """Sets the Optuna tags for the current MLFlow run
+
+        Args:
+            trial: Trial to be tracked
+            study: Study to be tracked
+        """
+        tags: Dict[str, str] = {}
+        tags["number"] = str(trial.number)
+        tags["datetime_start"] = str(trial.datetime_start)
+
+        tags["datetime_complete"] = str(trial.datetime_complete)
+
+        # Set state and convert it to str and remove the common prefix.
+        trial_state = trial.state
+        if isinstance(trial_state, TrialState) and trial_state.is_finished():
+            tags["state"] = str(trial_state).split(".")[-1]
+
+        # Set direction and convert it to str and remove the common prefix.
+        study_direction = study.direction
+        if isinstance(study_direction, StudyDirection):
+            tags["direction"] = str(study_direction).split(".")[-1]
+
+        tags.update(trial.user_attrs)
+        tags.pop(RUN_ID_ATTRIBUTE_KEY, None)
+        distributions = {(k + "_distribution"): str(v) for (k, v) in trial.distributions.items()}
+        tags.update(distributions)
+
+        if self._tag_study_user_attrs:
+            tags.update(study.user_attrs)
+
+        # This is a temporary fix on Optuna side. It avoids an error with user
+        # attributes that are too long. It should be fixed on MLflow side later.
+        # When it is fixed on MLflow side this codeblock can be removed.
+        # see https://github.com/optuna/optuna/issues/1340
+        # see https://github.com/mlflow/mlflow/issues/2931
+        max_mlflow_tag_length = 5000
+        for key, value in tags.items():
+            value = str(value)  # make sure it is a string
+            if len(value) > max_mlflow_tag_length:
+                tags[key] = textwrap.shorten(value, max_mlflow_tag_length)
```

Comment:
```suggestion
        for key, value in tags.items():
            value = str(value)  # make sure it is a string
            if len(value) > max_mlflow_tag_length:
                tags[key] = textwrap.shorten(value, mlflow.utils.validation.MAX_TAG_VAL_LENGTH)
```

Reply:
thanks @harupy I integrated this suggestion.